test/libsolidity/SyntaxTest.cpp

```
@@ -46,15 +46,26 @@ SyntaxTest::SyntaxTest(
 	CommonSyntaxTest(_filename, _evmVersion),
 	m_minSeverity(_minSeverity)
 {
-	static std::set<std::string> const compileViaYulAllowedValues{"true", "false"};
-
 	auto const eofEnabled = solidity::test::CommonOptions::get().eofVersion().has_value();
 
-	m_compileViaYul = m_reader.stringSetting("compileViaYul", eofEnabled ? "true" : "false");
-	if (!util::contains(compileViaYulAllowedValues, m_compileViaYul))
-		BOOST_THROW_EXCEPTION(std::runtime_error("Invalid compileViaYul value: " + m_compileViaYul + "."));
+	m_compileViaYul = m_reader.enumSetting<CompileViaYul>(
+		"compileViaYul",
+		{
+			{toString(CompileViaYul::False), CompileViaYul::False},
+			{toString(CompileViaYul::True), CompileViaYul::True},
+			{toString(CompileViaYul::Also), CompileViaYul::Also},
+			{toString(CompileViaYul::OnlyOnEOF), CompileViaYul::OnlyOnEOF}
+		},
+		toString(CompileViaYul::OnlyOnEOF)
+	);
+
+	solUnimplementedAssert(m_compileViaYul != CompileViaYul::Also,
+		"'compileViaYul: also' is not yet supported in syntax tests.");
 
-	if (m_compileViaYul == "false" && eofEnabled)
+	if (bytecodeFormat().contains(BytecodeFormat::EOFv1) && m_compileViaYul == CompileViaYul::False)
+		BOOST_THROW_EXCEPTION(std::runtime_error("Compilation to EOF requires using Yul IR"));
+
+	if (m_compileViaYul == CompileViaYul::False && eofEnabled)
 		m_shouldRun = false;
```

Regarding your question on the chat, if I understand correctly, the problem now is that if we use `">=EOFv1,legacy"` as the default for `bytecodeFormat`, tests that have `compileViaYul: false` start failing the check above?

That's actually a good point, I did not consider it, but it's fortunately easy to solve. `false` used to be the default and back then we'd have to deal with thousands of tests, but now that we changed it to `onlyOnEOF` it's just ~40 tests that set `compileViaYul: false`. My suggestion would be to look at those tests and do one of two things:
- Switch from `false` to `onlyOnEOF`.
- Add `bytecodeFormat: legacy`.

I suspect that the second one will be appropriate in almost all cases because these tests have good reasons not to want an IR run at all (either have a separate IR version with different expectations or use non-IR features like abicoder v1), but you may find some where the first one is more appropriate.

---

A nice bonus will be that we'll also be able to make `bytecodeFormat: legacy` mandatory when `compileViaYul: false` is used. I.e. you can report an error in the `m_compileViaYul == CompileViaYul::False && eofEnabled` check above instead of skipping the run.

---

This is already handled by the condition 2 lines above.

---

Ah, you mean that the `contains(EOFv1)` check will never let us even reach the `eofEnabled` check? True. But we should still make it an assert.